File: packages/health_connector_hc_android/test/unit_tests/src/mappers/health_record_mappers/sleep/sleep_session_record_mapper_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:health_connector_core/health_connector_core_internal.dart';
import 'package:health_connector_hc_android/src/mappers/health_record_mappers/sleep/sleep_session_record_mapper.dart';
import 'package:health_connector_hc_android/src/pigeon/health_connector_hc_android_api.g.dart';

import '../../../../utils/fake_data.dart';

void main() {
  final fakeStage1StartTime = FakeData.fakeStartTime.add(
    const Duration(hours: 1),
  );
  final fakeStage1EndTime = FakeData.fakeStartTime.add(
    const Duration(hours: 3),
  );

  group(
    'SleepSessionRecordMapper',
    () {
      group(
        'SleepSessionRecordToDto',
        () {
          test(
            'converts SleepSessionRecord to SleepSessionRecordDto',
            () {
              const title = 'Night Sleep';
              const notes = 'Good rest';

              final record = SleepSessionRecord(
                id: HealthRecordId(FakeData.fakeId),
                startTime: FakeData.fakeStartTime,
                endTime: FakeData.fakeEndTime,
                startZoneOffsetSeconds:
                    FakeData.fakeStartTime.timeZoneOffset.inSeconds,
                endZoneOffsetSeconds:
                    FakeData.fakeEndTime.timeZoneOffset.inSeconds,
                metadata: Metadata.internal(
                  dataOrigin: const DataOrigin(FakeData.fakeDataOrigin),
                  recordingMethod: RecordingMethod.activelyRecorded,
                  clientRecordVersion: 1,
                  device: const Device(type: DeviceType.watch),
                ),
                samples: [
                  SleepStageSample(
                    startTime: fakeStage1StartTime,
                    endTime: fakeStage1EndTime,
                    stageType: SleepStage.deep,
                  ),
                ],
                title: title,
                notes: notes,
              );

              final dto = record.toDto();

              expect(dto.id, FakeData.fakeId);
              expect(
                dto.startTime,
                FakeData.fakeStartTime.millisecondsSinceEpoch,
              );
              expect(dto.endTime, FakeData.fakeEndTime.millisecondsSinceEpoch);
              expect(dto.metadata.dataOrigin, FakeData.fakeDataOrigin);
              expect(dto.stages, hasLength(1));
              expect(dto.stages[0].stage, SleepStageDto.deep);
              expect(dto.title, title);
              expect(dto.notes, notes);
            },
          );
        },
      );

      group(
        'SleepSessionRecordDtoToDomain',
        () {
          test(
            'converts SleepSessionRecordDto to SleepSessionRecord',
            () {
              const title = 'Afternoon Nap';

              final dto = SleepSessionRecordDto(
                id: FakeData.fakeId,
                startTime: FakeData.fakeLocalStartTime.millisecondsSinceEpoch,
                endTime: FakeData.fakeLocalEndTime.millisecondsSinceEpoch,
                startZoneOffsetSeconds:
                    FakeData.fakeStartTime.timeZoneOffset.inSeconds,
                endZoneOffsetSeconds:
                    FakeData.fakeEndTime.timeZoneOffset.inSeconds,
                metadata: MetadataDto(
                  dataOrigin: FakeData.fakeDataOrigin,
                  recordingMethod: RecordingMethodDto.manualEntry,
                  clientRecordVersion: 1,
                  deviceType: DeviceTypeDto.phone,
                ),
                stages: [
                  SleepStageSampleDto(
                    startTime: fakeStage1StartTime.millisecondsSinceEpoch,
                    endTime: fakeStage1EndTime.millisecondsSinceEpoch,
                    stage: SleepStageDto.rem,
                  ),
                ],
                title: title,
              );

              final record = dto.toDomain();

              expect(record.id.value, FakeData.fakeId);
              expect(record.startTime, FakeData.fakeStartTime);
              expect(record.endTime, FakeData.fakeEndTime);
              expect(
                record.metadata.dataOrigin?.packageName,
                FakeData.fakeDataOrigin,
              );
              expect(record.samples, hasLength(1));
              expect(record.samples[0].stageType, SleepStage.rem);
              expect(record.title, title);
            },
          );

          test(
            'converts SleepSessionRecordDto with null id to '
            'domain with none id',
            () {
              final dto = SleepSessionRecordDto(
                startTime: FakeData.fakeLocalStartTime.millisecondsSinceEpoch,
                endTime: FakeData.fakeLocalEndTime.millisecondsSinceEpoch,
                startZoneOffsetSeconds: 0,
                endZoneOffsetSeconds: 0,
                metadata: MetadataDto(
                  dataOrigin: FakeData.fakeDataOrigin,
                  recordingMethod: RecordingMethodDto.activelyRecorded,
                  clientRecordVersion: 1,
                  deviceType: DeviceTypeDto.phone,
                ),
                stages: [],
              );

              final record = dto.toDomain();

              expect(record.id, HealthRecordId.none);
            },
          );
        },
      );
    },
  );
}

File: packages/health_connector_hc_android/android/src/test/kotlin/com/phamtunglam/health_connector_hc_android/unit_tests/mappers/health_record_mappers/SleepSessionRecordMapperTest.kt
package com.phamtunglam.health_connector_hc_android.unit_tests.mappers.health_record_mappers

import androidx.health.connect.client.records.SleepSessionRecord
import androidx.health.connect.client.records.metadata.Metadata
import com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers.toDto
import com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers.toHealthConnect
import com.phamtunglam.health_connector_hc_android.pigeon.DeviceTypeDto
import com.phamtunglam.health_connector_hc_android.pigeon.MetadataDto
import com.phamtunglam.health_connector_hc_android.pigeon.RecordingMethodDto
import com.phamtunglam.health_connector_hc_android.pigeon.SleepSessionRecordDto
import com.phamtunglam.health_connector_hc_android.pigeon.SleepStageDto
import com.phamtunglam.health_connector_hc_android.pigeon.SleepStageSampleDto
import io.kotest.matchers.shouldBe
import java.time.Instant
import java.time.ZoneOffset
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test

@DisplayName("SleepSessionRecordMapper")
class SleepSessionRecordMapperTest {

    private companion object {
        const val TEST_ID = "sleep-id"
        const val TEST_START_TIME = 1609459200000L
        const val TEST_END_TIME = 1609489200000L // +8 hours approx
        const val TEST_STAGE_START = 1609460000000L
        const val TEST_STAGE_END = 1609461000000L
        const val TEST_TITLE = "Night Sleep"
        const val TEST_NOTES = "Good sleep"
        val TEST_ZONE_OFFSET: ZoneOffset = ZoneOffset.ofHours(1)
    }

    @Test
    @DisplayName("GIVEN SleepSessionRecord → WHEN toDto → THEN converts correctly")
    fun whenRecordToDto_thenConvertsCorrectly() {
        // Given
        val record = SleepSessionRecord(
            startTime = Instant.ofEpochMilli(TEST_START_TIME),
            endTime = Instant.ofEpochMilli(TEST_END_TIME),
            startZoneOffset = TEST_ZONE_OFFSET,
            endZoneOffset = TEST_ZONE_OFFSET,
            metadata = Metadata.manualEntry(),
            title = TEST_TITLE,
            notes = TEST_NOTES,
            stages = listOf(
                SleepSessionRecord.Stage(
                    startTime = Instant.ofEpochMilli(TEST_STAGE_START),
                    endTime = Instant.ofEpochMilli(TEST_STAGE_END),
                    stage = SleepSessionRecord.STAGE_TYPE_DEEP,
                ),
            ),
        )

        // When
        val result = record.toDto()

        // Then
        result.startTime shouldBe TEST_START_TIME
        result.endTime shouldBe TEST_END_TIME
        result.startZoneOffsetSeconds shouldBe TEST_ZONE_OFFSET.totalSeconds.toLong()
        result.endZoneOffsetSeconds shouldBe TEST_ZONE_OFFSET.totalSeconds.toLong()
        result.title shouldBe TEST_TITLE
        result.notes shouldBe TEST_NOTES
        result.stages.size shouldBe 1
        result.stages[0].stage shouldBe SleepStageDto.DEEP
        result.stages[0].startTime shouldBe TEST_STAGE_START
        result.stages[0].endTime shouldBe TEST_STAGE_END
    }

    @Test
    @DisplayName("GIVEN SleepSessionRecordDto → WHEN toHealthConnect → THEN creates correctly")
    fun whenDtoToHealthConnect_thenCreatesCorrectly() {
        // Given
        val dto = SleepSessionRecordDto(
            id = TEST_ID,
            startTime = TEST_START_TIME,
            endTime = TEST_END_TIME,
            startZoneOffsetSeconds = TEST_ZONE_OFFSET.totalSeconds.toLong(),
            endZoneOffsetSeconds = TEST_ZONE_OFFSET.totalSeconds.toLong(),
            title = TEST_TITLE,
            notes = TEST_NOTES,
            metadata = MetadataDto(
                dataOrigin = "com.example.app",
                recordingMethod = RecordingMethodDto.MANUAL_ENTRY,
                deviceType = DeviceTypeDto.PHONE,
            ),
            stages = listOf(
                SleepStageSampleDto(
                    startTime = TEST_STAGE_START,
                    endTime = TEST_STAGE_END,
                    stage = SleepStageDto.LIGHT,
                ),
            ),
        )

        // When
        val result = dto.toHealthConnect()

        // Then
        result.startTime shouldBe Instant.ofEpochMilli(TEST_START_TIME)
        result.endTime shouldBe Instant.ofEpochMilli(TEST_END_TIME)
        result.startZoneOffset shouldBe TEST_ZONE_OFFSET
        result.endZoneOffset shouldBe TEST_ZONE_OFFSET
        result.title shouldBe TEST_TITLE
        result.notes shouldBe TEST_NOTES
        result.stages.size shouldBe 1
        result.stages[0].stage shouldBe SleepSessionRecord.STAGE_TYPE_LIGHT
        result.stages[0].startTime shouldBe Instant.ofEpochMilli(TEST_STAGE_START)
        result.stages[0].endTime shouldBe Instant.ofEpochMilli(TEST_STAGE_END)
        result.metadata.id shouldBe TEST_ID
    }
}
